### PR TITLE
Add support for registering PDU allocators for TCP and UDP

### PR DESCRIPTION
**Pull request by ebertland**
Inner PDU protocols can be registered for well-known port numbers and associated with the source or destination port fields.

### CODE REVIEW

**Comment by ebertland** [comment]
I would like to upstream this enhancement.

File: src/tcp.cpp [hunk 1]
```diff
@@ -120,9 +121,21 @@ TCP::TCP(const uint8_t* buffer, uint32_t total_sz) {
...
+    
     // If we still have any bytes left
     if (stream) {
-        inner_pdu(new RawPDU(stream.pointer(), stream.size()));
+        PDU* new_pdu = nullptr;
+        if ((new_pdu = Internals::allocate<TCP>(
+                 {Allocators::SRC_PORT, sport()}, stream.pointer(), stream.size()))) {
+        }
+        else if ((new_pdu = Internals::allocate<TCP>(
+                      {Allocators::DST_PORT, dport()}, stream.pointer(), stream.size()))) {
```
Comment (jamesm-cerebras) [inline src/tcp.cpp:132]: Nit: alignment should match line 129 (also in udp.cpp)

Comment (ebertland) [reply src/tcp.cpp:132]: I agree that those should line up, but this is how emacs indents it -- based on the open parenthesis on the previous line, maybe? I created a config specifically for this library to match the upstream author's style. It's been good so far, so I'd rather not adjust this line.

**Comment by ebertland** [comment]
Please approve if you don't have any other comments, @jamesm-cerebras .

**Review by jamesm-cerebras — APPROVED** [review]

**Pull request merged**